From babe3d8cfbf85c23a71cb167008065bc63cc2fb0 Mon Sep 17 00:00:00 2001
Message-Id: <babe3d8cfbf85c23a71cb167008065bc63cc2fb0@dist-git>
From: Michal Privoznik <mprivozn@redhat.com>
Date: Mon, 4 Feb 2019 10:37:50 +0100
Subject: [PATCH] virfile: Detect ceph as shared FS

RHEL-7.7: https://bugzilla.redhat.com/show_bug.cgi?id=1665553
RHEL-7.6.z: https://bugzilla.redhat.com/show_bug.cgi?id=1672178

Ceph can be mounted just like any other filesystem and in fact is
a shared and cluster filesystem. The filesystem magic constant
was taken from kernel sources as it is not in magic.h yet.

Signed-off-by: Michal Privoznik <mprivozn@redhat.com>
Reviewed-by: Erik Skultety <eskultet@redhat.com>
(cherry picked from commit 6dd2a2ae6386b1d51edcc9a434f56d7f9dc2cb35)
Signed-off-by: Michal Privoznik <mprivozn@redhat.com>
Message-Id: <60cef73c5369ea9b37bf0663ee7f06e2605abd46.1549273057.git.mprivozn@redhat.com>
Reviewed-by: Erik Skultety <eskultet@redhat.com>
---
 src/util/virfile.c            | 9 ++++++++-
 src/util/virfile.h            | 1 +
 src/util/virstoragefile.c     | 3 ++-
 tests/virfiledata/mounts3.txt | 2 ++
 tests/virfilemock.c           | 5 +++++
 tests/virfiletest.c           | 2 ++
 6 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/src/util/virfile.c b/src/util/virfile.c
index 716b55d770..471d309062 100644
--- a/src/util/virfile.c
+++ b/src/util/virfile.c
@@ -3537,6 +3537,9 @@ int virFilePrintf(FILE *fp, const char *msg, ...)
 # ifndef FUSE_SUPER_MAGIC
 #  define FUSE_SUPER_MAGIC 0x65735546
 # endif
+# ifndef CEPH_SUPER_MAGIC
+#  define CEPH_SUPER_MAGIC 0x00C36400
+# endif
 
 # define PROC_MOUNTS "/proc/mounts"
 
@@ -3682,6 +3685,9 @@ virFileIsSharedFSType(const char *path,
     if ((fstypes & VIR_FILE_SHFS_CIFS) &&
         (f_type == CIFS_SUPER_MAGIC))
         return 1;
+    if ((fstypes & VIR_FILE_SHFS_CEPH) &&
+        (f_type == CEPH_SUPER_MAGIC))
+        return 1;
 
     return 0;
 }
@@ -3845,7 +3851,8 @@ int virFileIsSharedFS(const char *path)
                                  VIR_FILE_SHFS_OCFS |
                                  VIR_FILE_SHFS_AFS |
                                  VIR_FILE_SHFS_SMB |
-                                 VIR_FILE_SHFS_CIFS);
+                                 VIR_FILE_SHFS_CIFS |
+                                 VIR_FILE_SHFS_CEPH);
 }
 
 
diff --git a/src/util/virfile.h b/src/util/virfile.h
index 6f1e802fde..1d16e96b59 100644
--- a/src/util/virfile.h
+++ b/src/util/virfile.h
@@ -205,6 +205,7 @@ enum {
     VIR_FILE_SHFS_AFS = (1 << 3),
     VIR_FILE_SHFS_SMB = (1 << 4),
     VIR_FILE_SHFS_CIFS = (1 << 5),
+    VIR_FILE_SHFS_CEPH = (1 << 6),
 };
 
 int virFileIsSharedFSType(const char *path, int fstypes) ATTRIBUTE_NONNULL(1);
diff --git a/src/util/virstoragefile.c b/src/util/virstoragefile.c
index 58f67278da..0b15219c3e 100644
--- a/src/util/virstoragefile.c
+++ b/src/util/virstoragefile.c
@@ -1366,7 +1366,8 @@ int virStorageFileIsClusterFS(const char *path)
      */
     return virFileIsSharedFSType(path,
                                  VIR_FILE_SHFS_GFS2 |
-                                 VIR_FILE_SHFS_OCFS);
+                                 VIR_FILE_SHFS_OCFS |
+                                 VIR_FILE_SHFS_CEPH);
 }
 
 #ifdef LVS
diff --git a/tests/virfiledata/mounts3.txt b/tests/virfiledata/mounts3.txt
index 134c6e8f81..68eded048c 100644
--- a/tests/virfiledata/mounts3.txt
+++ b/tests/virfiledata/mounts3.txt
@@ -33,3 +33,5 @@ host:/nfs /nfs nfs4 rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,
 dev /nfs/blah devtmpfs rw,nosuid,relatime,size=10240k,nr_inodes=4093060,mode=755 0 0
 host:/gv0 /gluster fuse.glusterfs rw 0 0
 root@host:/tmp/mkdir /gluster/sshfs fuse.sshfs rw 0 0
+192.168.0.1:/ceph/data /ceph ceph rw,noatime,name=cephfs,secret=<hidden>,acl,wsize=16777216 0 0
+192.168.0.1,192.168.0.2,192.168.0.3:/ceph/data2 /ceph/multi ceph rw,noatime,name=cephfs,secret=<hidden>,acl,wsize=16777216 0 0
diff --git a/tests/virfilemock.c b/tests/virfilemock.c
index ae5c8d025a..eb5182df66 100644
--- a/tests/virfilemock.c
+++ b/tests/virfilemock.c
@@ -88,6 +88,9 @@ setmntent(const char *filename, const char *type)
 #ifndef FUSE_SUPER_MAGIC
 # define FUSE_SUPER_MAGIC 0x65735546
 #endif
+#ifndef CEPH_SUPER_MAGIC
+# define CEPH_SUPER_MAGIC 0x00c36400
+#endif
 
 
 static int
@@ -134,6 +137,8 @@ statfs_mock(const char *mtab,
             ftype = CIFS_SUPER_MAGIC;
         } else if (STRPREFIX(mb.mnt_type, "fuse")) {
             ftype = FUSE_SUPER_MAGIC;
+        } else if (STRPREFIX(mb.mnt_type, "ceph")) {
+            ftype = CEPH_SUPER_MAGIC;
         } else {
             /* Everything else is EXT4. We don't care really for other paths. */
             ftype = EXT4_SUPER_MAGIC;
diff --git a/tests/virfiletest.c b/tests/virfiletest.c
index a246d601ba..972c07fdc5 100644
--- a/tests/virfiletest.c
+++ b/tests/virfiletest.c
@@ -458,6 +458,8 @@ mymain(void)
     DO_TEST_FILE_IS_SHARED_FS_TYPE("mounts3.txt", "/gluster/file", true);
     DO_TEST_FILE_IS_SHARED_FS_TYPE("mounts3.txt", "/gluster/sshfs/file", false);
     DO_TEST_FILE_IS_SHARED_FS_TYPE("mounts3.txt", "/some/symlink/file", true);
+    DO_TEST_FILE_IS_SHARED_FS_TYPE("mounts3.txt", "/ceph/file", true);
+    DO_TEST_FILE_IS_SHARED_FS_TYPE("mounts3.txt", "/ceph/multi/file", true);
 
     return ret != 0 ? EXIT_FAILURE : EXIT_SUCCESS;
 }
-- 
2.20.1

